File: core/domain/src/commonTest/kotlin/com/syrmos/core/domain/journey/ConnectionRiskTest.kt
package com.syrmos.core.domain.journey

import com.syrmos.core.model.journey.Feasibility
import com.syrmos.core.model.journey.FeasibilityStatus
import com.syrmos.core.model.journey.JourneyOption
import com.syrmos.core.model.journey.Leg
import com.syrmos.core.model.journey.LegKind
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDate
import kotlin.time.Duration.Companion.seconds
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * Kotlin peer of web `web-tests/connection-risk.test.js`, mirroring the golden
 * cases in fixtures/journeys/connection-risk.json. Proves S07 risk numbers +
 * status are identical on Kotlin and web (and, later, iOS). Cases inlined
 * (commonTest has no file IO); the JSON fixture is the shared source of truth.
 */
class ConnectionRiskTest {
    private val date = LocalDate.parse("2026-01-15")
    private fun inst(s: String) = Instant.parse(s)

    private fun ride(id: String, line: String, from: String, to: String,
                     dep: String? = null, arr: String? = null, unc: Int? = null) =
        Leg(id = id, kind = LegKind.RIDE, fromId = from, toId = to, lineId = line,
            departureInstant = dep?.let(::inst), arrivalInstant = arr?.let(::inst),
            serviceDate = date, uncertaintySeconds = unc)

    private fun transfer(id: String, from: String, to: String, min: Int?) =
        Leg(id = id, kind = LegKind.TRANSFER, fromId = from, toId = to, serviceDate = date,
            transferMinimumSeconds = min)

    private fun option(vararg legs: Leg) = JourneyOption(
        id = "o", requestId = "r", legs = legs.toList(),
        transferCount = legs.count { it.kind == LegKind.RIDE } - 1,
        feasibility = Feasibility(FeasibilityStatus.UNKNOWN, explanationCode = "pending"),
    )

    @Test
    fun tightTransfer() {
        val opt = option(
            ride("r0", "M1", "PIR", "OMO", arr = "2026-01-15T08:15:00+02:00"),
            transfer("t1", "OMO", "OMO3", 120),
            ride("r2", "M2", "OMO3", "SYN", dep = "2026-01-15T08:18:00+02:00"),
        )
        val rs = ConnectionRisk.risks(opt)
        assertEquals(1, rs.size)
        assertEquals("OMO", rs[0].atStationId)
        assertEquals("M2", rs[0].toLineId)
        assertEquals(FeasibilityStatus.TIGHT, rs[0].status)
        assertEquals(180, rs[0].availableSeconds)
        assertEquals(120, rs[0].recommendedSeconds)
        assertEquals(FeasibilityStatus.TIGHT, ConnectionRisk.worstStatus(opt))
    }

    @Test
    fun missedMatchesSpecExample() {
        val opt = option(
            ride("r0", "M1", "PIR", "OMO", arr = "2026-01-15T08:15:00+02:00"),
            transfer("t1", "OMO", "OMO3", 300),
            ride("r2", "M2", "OMO3", "SYN", dep = "2026-01-15T08:17:00+02:00"),
        )
        val r = ConnectionRisk.risks(opt).single()
        assertEquals(FeasibilityStatus.MISSED, r.status)
        assertEquals(120, r.availableSeconds) // "2 min available"
        assertEquals(300, r.recommendedSeconds) // "allow 5 min"
    }

    @Test
    fun twoTransfersWorstWins() {
        val opt = option(
            ride("r0", "M1", "PIR", "OMO", arr = "2026-01-15T08:15:00+02:00"),
            transfer("t1", "OMO", "OMO2", 120),
            ride("r2", "M2", "OMO2", "SYN", dep = "2026-01-15T08:25:00+02:00", arr = "2026-01-15T08:35:00+02:00"),
            transfer("t3", "SYN", "SYN3", 120),
            ride("r4", "M3", "SYN3", "AIR", dep = "2026-01-15T08:36:00+02:00"),
        )
        val rs = ConnectionRisk.risks(opt)
        assertEquals(listOf(FeasibilityStatus.COMFORTABLE, FeasibilityStatus.MISSED), rs.map { it.status })
        assertEquals(FeasibilityStatus.MISSED, ConnectionRisk.worstStatus(opt))
        val concern = ConnectionRisk.nextConcern(opt, 0)
        assertEquals("M3", concern?.toLineId)
        assertNull(ConnectionRisk.nextConcern(opt, 5))
    }

    @Test
    fun unknownWhenDepartureMissing() {
        val opt = option(
            ride("r0", "M1", "PIR", "OMO", arr = "2026-01-15T08:15:00+02:00"),
            transfer("t1", "OMO", "OMO3", 120),
            ride("r2", "M2", "OMO3", "SYN", dep = null),
        )
        val r = ConnectionRisk.risks(opt).single()
        assertEquals(FeasibilityStatus.UNKNOWN, r.status)
        assertNull(r.availableSeconds)
        assertEquals(120, r.recommendedSeconds)
    }

    @Test
    fun directHasNoRisks() {
        val opt = option(
            ride("r0", "M2", "SYN", "ELL", dep = "2026-01-15T08:00:00+02:00", arr = "2026-01-15T08:20:00+02:00"),
        )
        assertEquals(emptyList(), ConnectionRisk.risks(opt))
        assertEquals(FeasibilityStatus.COMFORTABLE, ConnectionRisk.worstStatus(opt))
        assertNull(ConnectionRisk.nextConcern(opt, 0))
    }

    @Test
    fun tightBoundaryInclusiveAt179() {
        fun statusForGap(gap: Int): FeasibilityStatus {
            val dep = Instant.parse("2026-01-15T08:00:00+02:00") + gap.seconds
            val opt = option(
                ride("r0", "M1", "A", "B", arr = "2026-01-15T08:00:00+02:00"),
                transfer("t1", "B", "B2", 120),
                ride("r2", "M2", "B2", "C", dep = dep.toString()),
            )
            return ConnectionRisk.risks(opt).single().status
        }
        assertEquals(FeasibilityStatus.TIGHT, statusForGap(299))       // margin 179
        assertEquals(FeasibilityStatus.COMFORTABLE, statusForGap(300)) // margin 180
        assertEquals(FeasibilityStatus.MISSED, statusForGap(119))      // margin -1
    }
}

File: web-tests/connection-risk.test.js
'use strict';
// 3.0 S07: the web connection-risk transform must match the shared golden fixture
// fixtures/journeys/connection-risk.json exactly — the same per-transfer rows and
// worst status Kotlin ConnectionRisk and iOS ConnectionRisk produce. Cross-client
// parity guard for the S07 risk numbers.
const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const RES = path.join(__dirname, '..', 'composeApp', 'src', 'wasmJsMain', 'resources');
const FIX = path.join(__dirname, '..', 'fixtures', 'journeys', 'connection-risk.json');
const Risk = require(path.join(RES, 'web-connection-risk.js'));
const fixture = JSON.parse(fs.readFileSync(FIX, 'utf8'));

test('every fixture case yields the exact expected per-transfer risks', () => {
  for (const c of fixture.cases) {
    assert.deepEqual(Risk.risks(c.option), c.expectedRisks, c.id);
  }
});

test('worst status per case matches the fixture', () => {
  for (const c of fixture.cases) {
    assert.equal(Risk.worstStatus(c.option), c.expectedWorstStatus, c.id);
  }
});

test('margin thresholds: tight boundary is inclusive at 179s', () => {
  // gap 299, allow 120 -> margin 179 -> tight; gap 300 -> margin 180 -> comfortable
  const at = (gap) => Risk.risks({ legs: [
    { id: 'r0', kind: 'ride', lineId: 'M1', fromId: 'A', toId: 'B', arrivalInstant: '2026-01-15T08:00:00+02:00' },
    { id: 't1', kind: 'transfer', fromId: 'B', toId: 'B2', transferMinimumSeconds: 120 },
    { id: 'r2', kind: 'ride', lineId: 'M2', fromId: 'B2', toId: 'C', departureInstant: new Date(Date.parse('2026-01-15T08:00:00+02:00') + gap * 1000).toISOString() },
  ] })[0].status;
  assert.equal(at(299), 'tight');
  assert.equal(at(300), 'comfortable');
  assert.equal(at(119), 'missed');
});

test('uncertainty erodes the margin', () => {
  const r = Risk.risks({ legs: [
    { id: 'r0', kind: 'ride', lineId: 'M1', fromId: 'A', toId: 'B', arrivalInstant: '2026-01-15T08:00:00+02:00', uncertaintySeconds: 120 },
    { id: 't1', kind: 'transfer', fromId: 'B', toId: 'B2', transferMinimumSeconds: 120 },
    { id: 'r2', kind: 'ride', lineId: 'M2', fromId: 'B2', toId: 'C', departureInstant: '2026-01-15T08:05:00+02:00' },
  ] })[0];
  // gap 300, allow 120, uncertainty 120 -> margin 60 -> tight (would be comfortable at uncertainty 0)
  assert.equal(r.availableSeconds, 300);
  assert.equal(r.status, 'tight');
});

test('nextConcern picks the nearest non-comfortable upcoming transfer, or null', () => {
  const opt = fixture.cases.find((c) => c.id === 'two_transfers_worst_wins').option;
  const concern = Risk.nextConcern(opt, 0);
  assert.equal(concern.toLineId, 'M3');
  assert.equal(concern.status, 'missed');
  // From beyond the last transfer there is nothing left to worry about.
  assert.equal(Risk.nextConcern(opt, 5), null);
  // A comfortable-only trip has no concern.
  assert.equal(Risk.nextConcern(fixture.cases.find((c) => c.id === 'comfortable_transfer').option, 0), null);
});
